// minecraft-net/index.js
var Socket = require(__dirname + '/lib/socket.js'),
	Server = require(__dirname + '/lib/server.js');

var connect = function(options, callback) {
	var socket = new Socket(options);
	socket.connect(options, callback);
	return socket;
};

var createServer = function(options, connectListener) {
	if(typeof options === 'function') {
		connectListener = options;
		options = null;
	}

	var server = new Server(options);
	server.on('connection', connectListener);
	return server;
};

module.exports = {
	constants: require(__dirname + '/lib/constants.js'),
	protocol: require(__dirname + '/lib/protocol.js'),
	Packet: require(__dirname + '/lib/packet.js'),
	Socket: Socket,
	ping: require(__dirname + '/lib/ping.js'),

	connect: connect,
	createConnection: connect,

	createServer: createServer,

	text: {
		black: 			'§0',
		darkBlue: 		'§1',
		darkGreen: 		'§2',
		darkAqua: 		'§3',
		darkRed: 		'§4',
		purple: 		'§5',
		gold: 			'§6',
		grey: 			'§7',
		darkGrey: 		'§8',
		indigo: 		'§9',
		green: 			'§a',
		aqua: 			'§b',
		red: 			'§c',
		pink: 			'§d',
		yellow: 		'§e',
		white: 			'§f',


		random: 		'§k',
		bold: 			'§l',
		strike: 		'§m',
		underline: 		'§n',
		italic: 		'§o',
		reset: 			'§r'
	}
};
